### DISCUSSION TITLE
revalidate not work in next13

### DISCUSSION THREAD

**Discussion by premier213** [vercel/next.js#42290] [Help] [answered]
I use revalidate mode not work to get data, this is an example in beta.nextjs.org.

```
export const revalidate = 30;

async function getData() {
  const res = await fetch('https://...', { next: { revalidate: 10 } });
  return res.json();
}

export default async function Page() {
  const data = await getData();
  // ...
}
```

**Comment by icyJoseph** [comment]
Are you using `next 13.0.1`? I do think there has been some fixes around revalidation.

**Reply by premier213** [reply]
I get the new update but that is not work

**Comment by vkatsou** [comment]
It's not working on next 13.0.2 either.
Tried `export const revalidate = 5;`, `next: { revalidate: 10},` and `cache: 'no-store',` but nothing seems to be working.

**Reply by leerob** [reply]
Can you share a reproducible example please? 🙏

**Reply by vkatsou** [reply]
Here is a really small (little messy) example.
Trying to revalidate every one second or revalidate each time with cache: no-store but unfortunately I can't.
https://github.com/vkatsou/next-13-revalidate-fetch-example

**Reply by leerob** [reply]
You're doing `cache: "no-store"` with revalidate. `no-store` is essentially SSR, so your fetch is going to run every time.

**Reply by leerob** [reply]
Also, your nested layout inside `home/layout` should _not_ include `html` and `body` tags. Only the root layout should.

**Reply by vkatsou** [reply]
> You're doing `cache: "no-store"` with revalidate. `no-store` is essentially SSR, so your fetch is going to run every time.

That's the problem, my fetch is not running every time. I've put all three possible ways in order to be easy to comment out each one of them.

**Reply by Alapottra-troj1n** [reply]
@vkatsou  were you able to fix it ?

**Reply by vkatsou** [reply]
Nope, but in the build is working as expected. Only on dev I face this problem.

**Reply by Gyurmatag** [reply]
This is still not working with 13.2.4 version. Not on dev, not on prod. The cache cannot be revalidated and cannot be set to no-store. Anyone got it working?

**Reply by jwalcher** [reply]
I have a workaround that clears the fetch-cache directory before calling `res.revalidate()` and works fine in latest 13.2.5 canary. See  https://github.com/vercel/next.js/issues/47273 for details

**Reply by Gyurmatag** [reply]
Thanks. I see the issue, but I don't want workarounds. As I see the issue was merged, hoping that we will get a release soon with that fix...

**Reply by jwalcher** [reply]
it's working now in 13.2.5-canary.26, at least in production. For dev mode, I'm still using my workaround :)

**Reply by Gyurmatag** [reply]
That's okay, but I don't want workarounds, I want it to work in a stable version flawlessly without a workaround. :)

**Comment by calitb** [comment]
not working for me on 13.0.5

**Comment by Basovs** [comment]
Next version 13.1.1 - Still not working. Noore "export const revalidate = 0" in page or layout components. Noore "cache: 'no-store'" in .fetch() option. Noore "{next:{revalidate: 0}}" in .fetch().

I do "yarn build" -> "yarn start" on the first initial load the data is SSR:ly fetched yes, but when I leave and come back to the same page/component the data ir returned from cache and not SSR:ly fetched anymore.... :(

**Reply by logemann** [reply]
I just did some tests with 13.1.1 and i have no problems. But i have not tested with revalidate=0 but played around with the different "cache" options and revalidate > 0. All these combinations work as expected when spinning up via **yarn start**

**Reply by Basovs** [reply]
@logemann I just noticed that indeed SSR works in all scenarious BUT ONLY when i manually refresh the page. If i navigate using the Link component from next/link then its not working... I do have spinner while fetching SSRrly the data. So i see that first entering the page (even using the Link component) the spinner is spinning/loading and thats a good sign. The second time when i enter the page with Link component it does not spin/load. Thats bad sign because that means the data come from cache and not fetching against the DB SSRly.

Some maybe good things to know. I do fetch against pages/api/products.ts API endpoint which then fetches against a swell.is store API using swells SDK.

**Reply by logemann** [reply]
You are right. It's only on refresh and not on next/link . This is not how it should work right? Nobody in the wild will do full refreshs.

**Comment by antonmoiseev** [comment]
That's probably the intended behavior. Trying to figure it out [here](https://github.com/vercel/next.js/discussions/44699) and there is the same question [here](https://github.com/vercel/next.js/discussions/41745#discussioncomment-4620262).

**Reply by logemann** [reply]
But then the whole revalidation is useless. New users will get the latest data anyway. And current users don't do refresh.

**Comment by oaj** [comment]
I have the same problem with 13.1.2
Also on demand revalidate does not work - `"await res.revalidate(url)"`

**Comment by westprophet** [comment]
With Next.js 13.1.6 revalidate does not work( 

But on local server with "next start" revalidation working good!

**Reply by Basovs** [reply]
v13.2.1 still doesnt work...
export const revalidate = 0 - doesnt work
export const dynamic = 'force-dynamic' - doesnt work
I think there is some kind of stupid caching going on... caching has never been good friend :)

**Comment by spigelli** [comment]
Running with force-static and revalidate appears to be working when running builds but not in dev mode. See https://github.com/spigelli/revalidation

**Reply by spigelli** [reply]
Auto and dynamic both default to no caching but I think that's the intended behavior?

**Comment by mdrayer** [comment]
I'm testing out `next@13.2.4-canary.1`. Using `revalidate` on the `fetch()` works fine for me both with `next dev` and with `next build` + `next start`. However, [on-demand revalidation](https://beta.nextjs.org/docs/data-fetching/revalidating#on-demand-revalidation) does not seem to be working via `res.revalidate()` on pages generated with `/app` directory when using `next build` + `next start`. It does seem to re-run the fetch, but the cache doesn't update.

On-demand revalidation _does_ still work on pages generated with `/pages` directory. 

It looks like `revalidatePath()` and `revalidateTag()` are in the works though: https://vercel.com/blog/vercel-cache-api-nextjs-cache#revalidating-components-with-cache-tags 🤞🤞

**Reply by DomVinyard** [reply]
I have the same experience, also in `next@13.2.4-canary.1`. On-demand revalidation is not working at all.

> It does seem to re-run the fetch, but the cache doesn't update.

Yes, this was my hypothesis too. 

Does anybody know if there's a roadmap for `revalidatePath()`? Or a workaround

**Reply by ghost** [reply]
Same issue here, revalidate works fine but on demand is not working in version 13.2.4. Revalidation is triggering but the api response is the cached variant and not the new update.

It would also be nice how we can rewrite the pages/api to app/api with a revalidate function. How does one achieve that?

**Reply by DomVinyard** [reply]
@ddegooijer the docs imply that is not possible yet https://beta.nextjs.org/docs/data-fetching/revalidating

<img width="763" alt="image" src="https://user-images.githubusercontent.com/1271197/224716178-b669809d-e4f4-4c8a-bc19-d375cac5acba.png">

**Reply by ghost** [reply]
> @ddegooijer the docs imply that is not possible yet https://beta.nextjs.org/docs/data-fetching/revalidating
> 
> <img alt="image" width="763" src="https://user-images.githubusercontent.com/1271197/224716178-b669809d-e4f4-4c8a-bc19-d375cac5acba.png">

Ahh oke, thanks for pointing that out. Hopefully the NEXTjs team can clear things up with the on-demand revalidation not working.

**Comment by tlgsn** [comment]
Has this issue been resolved?

**Comment by ghost** [comment]
Hi,

I've still got an issue with the latest stable version of NEXTjs and revalidating data. The data returning was always the old one from the first build. Fixed it for now by adding a date timestamp to the api / graphql url.

**Reply by smitbarmase** [reply]
Yup, not working in latest stable version.

**Comment by arturomaciasmx** [comment]


**Reply by Alapottra-troj1n** [reply]
development mode works. the problem is within deployed version

**Comment by Alapottra-troj1n** [comment]
Did anyone figured out how to fix the issue ?

**Reply by 8lane** [reply]
I'm having this issue still on latest `13.2.4`.

**Reply by Gyurmatag** [reply]
Yeah, mee too. Hope It will be fixed ASAP. Super annoying.

**Comment by karlhorky** [comment]
It sounds like some of you may also be having the issue with soft navigation that a number of community members are experiencing.

I would suggest that you add your feedback to the thread over here:

- #42991

**Comment by overpeaks** [comment]
I'm also having an issue with revalidating on the most recent stable version (13.2.4).
I've tried both:
`const res = await fetch(process.env.CMS_API + "/items/" + id, {
    next: { revalidate: 5 }
  });`
during fetches, and putting:
`export const revalidate = 5;`
in layout.tsx

When I run `npm run dev` it works fine and correctly pulls the latest data from the CMS. But if I do `npm run build` + `npm run start` it immediately goes back to the old data it has cached. 

The only current fix I've found (that works for me) is deleting the cache folder in .next dir and rerunning the build and start scripts. Kind of annoying if I have to delete the cache folder and rerun the build and start scripts every time I make a change in the CMS...

**Comment by ghost** [comment]
Yes this is still an issue, revalidation does not work on 13.2.4. It always returns the cached data from build time, not the revalidated data. Tried it both with graphQL and REST, no dice.

**Comment by jwalcher** [comment]
I filed a bug two weeks ago
https://github.com/vercel/next.js/issues/47273 upvote?
My workaround is to clear the cache from within the route handler before calling res.revalidate. Not pretty, but no performance loss for my use case. 
I have problems also in dev mode.

**Reply by ghost** [reply]
Do you have an example? Thanks!

**Reply by jwalcher** [reply]
i just pushed to
https://github.com/jwalcher/no-no-store/
what I could swear was working exactly as intended with one of the previous canary's a few days ago but not currently....
I am frankly surprised this is taking them so long to fix, although it certainly will be great once it's finished.

**Reply by jwalcher** [reply]
I was missing 
```
 experimental: {
    appDir: true,
    isrMemoryCacheSize: 50
  },
```
in `next.config.js`. now works again in latest canary

**Comment by Alapottra-troj1n** [comment]
Btw there's is another solution I came across is if you're using the api route handlers folder of the app directory. Try to use GET and POST in the same route. Like /api/posts if it's a GET route it should have a POST function in the same route.js else Next.js builds it as a static route. Then just when you call it make the revalidation accordingly. It should work

**Reply by jwalcher** [reply]
well, at the moment the `{ NextResponse } from next/server` api is missing a `revalidate` method, so we can't revalidate from app directory route handler (though I assume/hope this will be possible in the future).

**Comment by LiahMartens** [comment]
This seems like a real blocker for the app directory to be production ready. I got it to work by completely disabling cache but this is obviously not ideal and it increased the build time significantly. Hoping this issue gets resolved soon

**Comment by niemiaszek** [comment]
I'm learning nextjs and building my first app. For me on current 13.2.5-canary.30 I'm still not getting ISR on-demand revalidation (even with `isrMemoryCacheSize: 50`) :/ Looking forward for this to work in my [Repo](https://github.com/niemiaszek/exchange-office-webpage). Maybe i'm doing something wrong, as I can see @jwalcher getting it to work

**Reply by niemiaszek** [reply]
Seems like it indeed started working after update. Still working on another issue I got related to reading/writing to `/tmp/...` files [here](https://github.com/vercel/next.js/discussions/47988)

**Comment by mshez** [comment]
Nops. Data is fetched and cached. It never revalidates nor with 'no-store', { next:  { revalidate: 0} }. Only way is rebuilding the project. @leerob any comments ?

**Reply by jwalcher** [reply]
Try the latest version of https://github.com/jwalcher/no-no-store/
Visiting `api/revalidate` in production calls `res.revalidate` and refreshes the data exactly as described in the docs. In dev mode, we instead clear the `cache/fetch-cache`, and then refresh. (The great thing is that it works not just with `fetch()` API but also apollo client, and indeed all queries are deduped during build.) (The `isrMemoryCacheSize: 50` is still necessary for the workaround in dev mode, but not in production any longer.)

**Comment by Dupflo** [comment]
In my case, the fetch function is fired on revalidation (every 10 sec for the test) but it always returns a version of my data so no changed applied

`const res = await fetch(/api/products/${slug}, {
      next: { revalidate: 10 },
    }).then((res) => res.json());`

**Reply by zsrobinson** [reply]
Yeah, it seems like whenever I deploy the site to Vercel, it says that the entire site is static and doesn't see that there's a revalidate option in the fetch() call. It just fetches at build time instead of doing any sort of revalidating. I thought upgrading to v13.3.0 would fix this, but I'm still seeming to have the same issue.

**Reply by josephakinlade** [reply]
I have the same issues, I've used all the possible ways provided in the docs none seems to work

**Comment by taufik-nurhidayat** [comment]
I have same issue here, using nextjs 13.2.4 and appDir feature. So i use revalidate api for ISR, but it's not working. Nah, i try to redeploy it and for first time i open the page the data is newer bat after i reload the page data is using old again.

**Comment by BIhab** [comment]
Same here, with Next 13.3.0, it will be great if a small patch came with a solution to that issue

**Reply by taufik-nurhidayat** [reply]
hei, i upgrade it to the 13.3..0 version and thats work fine

**Reply by BIhab** [reply]
Solved in last update 13.3.1

**Comment by ryangwn** [comment]
In Next 13.3.0, it works fine.

**Comment by calvincchong** [comment]
I am on Next 13.1.6 - revalidate cache does not work when I am deploying on Vercel. I am not using an API route but rather a helper function that uses and ORM to retrieve data. This data won't get updated when I set 
export const revalidate = 10;

**Reply by BIhab** [reply]
Try to update to the latest version

**Reply by AlexAnschuetz** [reply]
it doesnt work for me in `13.3.1` either :/

**Comment by mbathie** [comment]
Not working for me either on 13.3.2

**Comment by akashKarmakar02** [comment]
can some one provide a way around it

**Comment by anthropitec** [comment]
+1 to this problem on "next": "^13.3.1" ⏳

**Comment by ttiras** [comment]
Works locally but not on vercel production build. "next": "13.4.1"

**Reply by josephakinlade** [reply]
I get the same issues too

**Comment by bitstorm-tech** [comment]
I really appreciate the work of the devs for this masterpiece of software! But the cache system is a crucial part that does not work properly at the moment. For me personally (and I think for a lot of others too), this is a real showstopper at the moment. 

So, I wonder why this problem gets so less attention? I don't have the feeling, that this problem will be fixed soon. And sadly, as far as I can see, there is also no workaround at the moment? I feel a bit helpless right now :(

Btw., I use Supabase for the backend, but I don't think that has a big influence on the problem.

**Reply by Basovs** [reply]
Exactly my thoughts...

**Reply by josephakinlade** [reply]
It's a pain in the a** as I need to deliver a project only to discover the caching system doesn't revalidate

**Comment by ghost** [comment]
This could use some more love from the devs.

**Reply by bitstorm-tech** [reply]
At least it would be good to know if this is really a bug or working as intended, so we can move on to another framework.

**Comment by leerob** [comment, marked as answer]
Hey folks – appreciate the feedback here, but without a minimal reproduction and following the issue template, we cannot investigate any of these reports. I am going to create an issue from this discussion and we can prioritize it when there is a minimal reproduction added that follows the issue template 🙏

**Reply by leerob** [reply]
https://github.com/vercel/next.js/issues/49417
